[1/N] [Dispatchable Collectives] Create Backend class

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #83916 [7/N] [Dispatchable Collectives] Update reduce with CPU / CUDA implementations
* #83876 [6/N] [Dispatchable Collectives] Update recv with CPU / CUDA implementations
* #83859 [5/N] [Dispatchable Collectives] Update send with CPU / CUDA implementations
* #83810 [4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations
* #83735 [3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations
* #83680 [2/N] [Dispatchable Collectives] Extract ProcessGroup::Work into a separate class and update references
* **#83679 [1/N] [Dispatchable Collectives] Create Backend class**

### Changes:
- Create a new Backend class which contains collectives similar to that of https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroup.hpp.

### Context

https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D38839213](https://our.internmc.facebook.com/intern/diff/D38839213)

H-Huang committed:
[1/N] [Dispatchable Collectives] Create Backend class

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38839213).

H-Huang committed:
Update on "[1/N] [Dispatchable Collectives] Create Backend class"


### Changes:
- Create a new Backend class which contains collectives similar to that of https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroup.hpp.

#### Motivation
In future PRs, the existing ProcessGroupNCCL/Gloo/UCC will be migrated to derive from this Backend class. The idea is that we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type.

Differential Revision: [D38839213](https://our.internmc.facebook.com/intern/diff/D38839213)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38839213).

H-Huang committed:
Update on "[1/N] [Dispatchable Collectives] Create Backend class"


### Changes:
- Create a new Backend class which contains collectives similar to that of https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroup.hpp.

#### Motivation
In future PRs, the existing ProcessGroupNCCL/Gloo/UCC will be migrated to derive from this Backend class. The idea is that we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type.

Differential Revision: [D38839213](https://our.internmc.facebook.com/intern/diff/D38839213)

[ghstack-poisoned]

vadimkantorov commented:
This would be nice because even in GPU training sometimes broadcasting CPU tensor is needed. And in reverse, for CPU-only inference, a different backend should be used...

kwen2501 commented on torch/csrc/distributed/c10d/Backend.hpp:
```diff
@@ -0,0 +1,276 @@
+            "Backend ",
+            backendName,
+            " does not yet support sequence numbers."));
+  }
```
Just for our record: we'd need to think whether `SequenceNumber` related methods should stay here or be moved to `ProcessGroup`. For the purpose of smooth refactorization, they can stay here for now.

H-Huang replied:
Makes sense, agree on this!

kwen2501 reviewed (approved):
LGTM! Thanks!

H-Huang committed:
Update on "[1/N] [Dispatchable Collectives] Create Backend class"


### Changes:
- Create a new Backend class which contains collectives similar to that of https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroup.hpp.

#### Motivation
In future PRs, the existing ProcessGroupNCCL/Gloo/UCC will be migrated to derive from this Backend class. The idea is that we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type.

Differential Revision: [D38839213](https://our.internmc.facebook.com/intern/diff/D38839213)

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
